src/analysis/seminorm.lean

```diff
@@ -347,25 +505,25 @@ by { use 0, rintro _ ⟨x, rfl⟩, exact add_nonneg (p.nonneg _) (q.nonneg _) }
 
 noncomputable instance : has_inf (seminorm 𝕜 E) :=
 { inf := λ p q,
-  { to_fun := λ x, ⨅ u : E, p u + q (x-u),
-    triangle' := λ x y, begin
-      refine le_cinfi_add_cinfi (λ u v, _),
-      apply cinfi_le_of_le (bdd_below_range_add _ _ _) (v+u), dsimp only,
-      convert add_le_add (p.triangle v u) (q.triangle (y-v) (x-u)) using 1,
-      { rw show x + y - (v + u) = y - v + (x - u), by abel },
-      { abel },
-    end,
-    smul' := λ a x, begin
-      obtain rfl | ha := eq_or_ne a 0,
-      { simp_rw [norm_zero, zero_mul, zero_smul, zero_sub, seminorm.neg],
-        refine cinfi_eq_of_forall_ge_of_forall_gt_exists_lt
-          (λ i, add_nonneg (p.nonneg _) (q.nonneg _))
-          (λ x hx, ⟨0, by rwa [map_zero, map_zero, add_zero]⟩) },
-      simp_rw [real.mul_infi_of_nonneg (norm_nonneg a), mul_add, ←p.smul, ←q.smul, smul_sub],
-      refine function.surjective.infi_congr ((•) a⁻¹ : E → E) (λ u, ⟨a • u, inv_smul_smul₀ ha u⟩)
-        (λ u, _),
-      rw smul_inv_smul₀ ha,
-    end } }
+begin
+  refine seminorm.of (λ x, ⨅ u : E, p u + q (x-u)) _ _,
+  { intros x y,
+    refine le_cinfi_add_cinfi (λ u v, _),
+    apply cinfi_le_of_le (bdd_below_range_add _ _ _) (v + u), dsimp only,
+    convert add_le_add (p.add_le v u) (q.add_le (y-v) (x-u)) using 1,
+    { rw show x + y - (v + u) = y - v + (x - u), by abel },
+    { abel }},
+  { intros a x,
+    obtain rfl | ha := eq_or_ne a 0,
+    { rw [norm_zero, zero_mul, zero_smul],
+      refine cinfi_eq_of_forall_ge_of_forall_gt_exists_lt
+        (λ i, add_nonneg (p.nonneg _) (q.nonneg _))
+        (λ x hx, ⟨0, by rwa [map_zero, sub_zero, map_zero, add_zero]⟩) },
+    simp_rw [real.mul_infi_of_nonneg (norm_nonneg a), mul_add, ←p.smul, ←q.smul, smul_sub],
+    refine function.surjective.infi_congr ((•) a⁻¹ : E → E) (λ u, ⟨a • u, inv_smul_smul₀ ha u⟩)
+      (λ u, _),
+    rw smul_inv_smul₀ ha },
+end }
```

Comment:
Can you define this for the additive case too? Does the same proof work?

Reply:
I haven't thought about this, but this definition uses substraction, so it would not work for `add_monoid_seminorm`.

Reply:
Sure it would, you just add a section with lemmas about `add_monoid_seminorm` when `add_comm_group E`

Reply:
I meant it would not work for `add_monoid E`.

Reply:
Ah, sorry; I'm suggesting you define it as `noncomputable instance [add_comm_group E] : has_inf (add_monoid_seminorm 𝕜 E) :=`, because then you can define the `seminorm` version in terms of the `add_monoid_seminorm` version.

Reply:
I think this doesn't work for `add_monoid_seminorm` (even for `add_comm_group E`). You can write the `has_inf` instance as in the seminorm case, but you cannot prove the `le_inf` property in the `lattice` instance (so it would not really be an `inf`): the proof for `seminorm`uses `le_insert'`, which relies on `lemma neg (x : E) : p (-x) = p x`. But this isn't true for all `add_monoid_seminorms` (for instance, consider `p : Z/(3) \to R` given by p(0)=0, p(1)=1, p(-1)=2).
(As an aside, if you require the `add_monoid_seminorm` to be ultrametric, i.e., you ask p(x - y) \le max {p (x), p(y)}, then the lemma `neg` holds).

Reply:
Is that `p` usually considered a valid seminorm?

Reply:
I don't know in anyone would be interested in seminorms for which p(-x) can be different than p(x) (I would guess not). I only care about ultrametric seminorms, so this never happens.

Reply:
I don't think so.

Reply:
Couldn't we consider adding the axiom:
```lean
(eq_of_add_eq_zero' : ∀ r s, r + s = 0 → to_fun r = to_fun s)
```
to `add_monoid_seminorm`?

Reply:
In fact, you can then probably define the `inf` as `⨅ (u v : E) (h : x = u + v), p u + q v`

Reply:
I went back to the previous `inf` definition and proved that with that we get a `lattice` in the `add_comm_group` case (and then reused this in the corresponding instance for `seminorm`).

Reply:
You're totally right about my definition being identically zero, I'm not sure how I missed that given I was speaking to Wrenna about something almost identical the other day.

Reply:
I think I agree that we don't need the add_monoid version; and even if we did, it would be much clearer to make that refactor in a separate PR.

So I'm 100% on board with making this PR add only `add_group.seminorm`

Reply:
The only reason I defined `add_monoid_seminorm`, instead of `add_group_seminorm`, is that the existing `seminorm` assumes the following:
```
structure seminorm (𝕜 : Type*) (E : Type*) [semi_normed_ring 𝕜] [add_monoid E] [has_scalar 𝕜 E] :=
```
If I only define `add_group_seminorm`, I cannot reuse it here.

Reply:
Should I replace the `[add_monoid E]` hypothesis in `seminorm` by `[add_group E]`?

Reply:
OK, sounds good to me. I've made this change and replaced `add_monoid_seminorm` by `add_group_seminorm` everywhere. I replaced the hypothesis `(eq_of_add_eq_zero' : ∀ r s, r + s = 0 → to_fun r = to_fun s)` by
`(neg' : ∀ r, to_fun (-r) = to_fun r)`.